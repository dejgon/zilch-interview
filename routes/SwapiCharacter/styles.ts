import { StyleSheet } from 'react-native';

import { Colors } from '../../consts';

export const styles = StyleSheet.create({
  errorText: {
    color: Colors.Red.RED_500,
    fontFamily: 'Oswald-Regular',
    fontSize: 18,
    fontStyle: 'normal',
    fontWeight: '400',
    lineHeight: 27,
  },
  mainView: {
    width: '100%',
    marginTop: 30,
  },
  personView: {
    flexDirection: 'column',
    gap: 20,
  },
});
