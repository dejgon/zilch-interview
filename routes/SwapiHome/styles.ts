import { StyleSheet } from 'react-native';

import { Colors } from '../../consts';

export const styles = StyleSheet.create({
  errorText: {
    color: Colors.Red.RED_500,
    fontFamily: 'Oswald-Regular',
    fontSize: 18,
    fontStyle: 'normal',
    fontWeight: '400',
    lineHeight: 27,
  },
  mainView: {
    width: '100%',
    marginTop: 30,
    gap: 15,
    flex: 1,
  },
  informationText: {
    fontFamily: 'Oswald-Regular',
    textAlign: 'center',
    fontSize: 16,
    fontStyle: 'normal',
    fontWeight: '400',
    lineHeight: 27,
    marginBottom: 10,
  },
  searchText: {
    fontFamily: 'Oswald-Light',
    textAlign: 'center',
    fontSize: 16,
    fontStyle: 'normal',
    fontWeight: '400',
    lineHeight: 27,
  },
  personNameText: {
    fontFamily: 'Oswald-Regular',
    fontSize: 18,
    fontStyle: 'normal',
    fontWeight: '400',
    lineHeight: 27,
    marginVertical: 10,
  },
  peopleFoundText: {
    fontFamily: 'Oswald-Regular',
    fontSize: 14,
    fontStyle: 'normal',
    fontWeight: '400',
  },
  seeMoreText: {
    fontFamily: 'Oswald-Regular',
    fontSize: 12,
    fontStyle: 'normal',
    fontWeight: '400',
  },
  personElement: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
});
