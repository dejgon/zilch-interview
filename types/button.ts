export enum ButtonType {
  Outlined = 'outlined',
  Contained = 'contained',
}
