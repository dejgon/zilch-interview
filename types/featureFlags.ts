export enum FeatureFlags {
  ApiAddress = 'api_address',
  ApiAddressAndroid = 'api_address_android',
  Maintenance = 'maintenance',
}
