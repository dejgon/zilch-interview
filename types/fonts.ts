export enum Fonts {
  OswaldRegular = 'Oswald-Regular',
  OswaldLight = 'Oswald-Light',
}

export enum FontWeight {
  Normal = '400',
  Bold = '600',
}
