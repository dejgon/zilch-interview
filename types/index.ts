export * from './button';
export * from './fonts';
export * from './rootStackParamList';
export * from './routes';
export * from './secureStoreKeys';
export * from './user';
export * from './featureFlags';
export * from './requestStatus';
export * from './person';
