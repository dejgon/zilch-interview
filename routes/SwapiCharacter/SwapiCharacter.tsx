import React from 'react';
import { Linking, View } from 'react-native';

import { NativeStackScreenProps } from '@react-navigation/native-stack';

import { Attribute } from './components';
import { styles } from './styles';
import { ButtonHighlight } from '../../components';
import { AuthLayout } from '../../layouts';
import { ButtonType, RootStackParamList, Routes } from '../../types';

type Props = NativeStackScreenProps<RootStackParamList, Routes.SwapiCharacter>;

export function SwapiCharacter(props: Props) {
  const {
    route: {
      params: { person },
    },
  } = props;

  const handleLink = async (link: string) => {
    const supported = await Linking.canOpenURL(link);

    if (supported) {
      await Linking.openURL(link);
    }
  };

  return (
    <AuthLayout>
      <View style={styles.mainView}>
        {person && (
          <View style={styles.personView}>
            <Attribute name="Name" value={person.name} />
            <Attribute name="Birth year" value={person.birth_year} />
            <Attribute name="Height" value={person.height} />
            <Attribute name="Hair color" value={person.hair_color} />
            <Attribute name="Skin color" value={person.skin_color} />
            <Attribute name="Eye color" value={person.eye_color} />
            <ButtonHighlight
              text="Check SWAPI documentation"
              type={ButtonType.Contained}
              onPress={() => handleLink(person.url)}
            />
          </View>
        )}
      </View>
    </AuthLayout>
  );
}
