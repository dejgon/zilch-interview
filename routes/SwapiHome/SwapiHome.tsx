import React, { useCallback, useState } from 'react';
import {
  ActivityIndicator,
  FlatList,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';

import { NativeStackScreenProps } from '@react-navigation/native-stack';
import axios from 'axios';

import { styles } from './styles';
import { Input } from '../../components';
import { BasicScreenLayout } from '../../layouts';
import { Person, RootStackParamList, Routes } from '../../types';

type Props = NativeStackScreenProps<RootStackParamList, Routes.Swapi>;

export function SwapiHome(props: Props) {
  const { navigation } = props;
  const [searchValue, setSearchValue] = useState<string>('');
  const [searchResult, setSearchResult] = useState<Person[]>();
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [error, setError] = useState<string | undefined>();
  let debounced: NodeJS.Timeout | number;

  async function searchForPeople(searchValue: string) {
    setIsLoading(true);
    try {
      const response = await axios.get(
        `https://swapi.dev/api/people?search=${searchValue}`
      );
      setSearchResult(response.data.results as Person[]);
    } catch (e) {
      console.error(e);
      setError('Error occurred, please try again');
    }
    setIsLoading(false);
  }

  const onSearchTextChange = useCallback((text: string) => {
    setSearchValue(text);
    debounce(() => searchForPeople(text));
  }, []);

  function debounce(func: Function) {
    clearTimeout(debounced);
    debounced = setTimeout(func, 1000);
  }

  return (
    <BasicScreenLayout>
      <View style={styles.mainView}>
        <Text style={styles.informationText}>
          Find your favourite Star Wars character
        </Text>
        <Input
          label="Search for the character"
          value={searchValue}
          onChangeText={onSearchTextChange}
        />
        {isLoading ? (
          <ActivityIndicator />
        ) : searchResult && searchResult.length > 0 ? (
          <>
            <Text style={styles.peopleFoundText}>
              People found: {searchResult.length}
            </Text>
            <FlatList
              showsVerticalScrollIndicator
              persistentScrollbar={true}
              data={searchResult}
              renderItem={({ item }) => (
                <TouchableOpacity
                  testID="search-result"
                  style={styles.personElement}
                  key={item.name}
                  onPress={() =>
                    navigation.navigate(Routes.SwapiCharacter, {
                      person: item,
                    })
                  }>
                  <Text style={styles.personNameText}>‚ö™Ô∏è {item.name}</Text>
                  <Text>‚û°Ô∏è</Text>
                </TouchableOpacity>
              )}
              keyExtractor={(item) => item.name}
            />
          </>
        ) : (
          <Text style={styles.searchText}>
            No results, try something else üîé
          </Text>
        )}
        {error && <Text style={styles.errorText}>{error}</Text>}
      </View>
    </BasicScreenLayout>
  );
}
