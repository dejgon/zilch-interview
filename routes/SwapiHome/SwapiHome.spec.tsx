import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import { fireEvent, render, waitFor } from '@testing-library/react-native';
import axios from 'axios';
import { Provider } from 'react-redux';

import { SwapiHome } from './SwapiHome';
import { swapiResponse } from '../../mocks';
import { store } from '../../store';
import { RootStackParamList, Routes } from '../../types';
import { SwapiCharacter } from '../SwapiCharacter';

const Stack = createNativeStackNavigator<RootStackParamList>();

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

function setup() {
  return render(
    <Provider store={store}>
      <NavigationContainer>
        <Stack.Navigator initialRouteName={Routes.Swapi}>
          <Stack.Screen
            name={Routes.Swapi}
            component={SwapiHome}
            options={{ headerShown: true }}
          />
          <Stack.Screen
            name={Routes.SwapiCharacter}
            component={SwapiCharacter}
            options={{ headerShown: true }}
          />
        </Stack.Navigator>
      </NavigationContainer>
    </Provider>
  );
}

describe('SwapiHome.tsx', () => {
  it('renders properly', () => {
    const component = setup();
    expect(component).toMatchSnapshot();
  });

  it('gets the data from API', async () => {
    mockedAxios.get.mockResolvedValueOnce({ data: swapiResponse });
    const { getByText, getByTestId, getByDisplayValue } = setup();

    expect(getByText('Search for the character')).toBeDefined();

    const input = getByTestId('input-text-area');

    expect(input).toBeDefined();

    fireEvent.changeText(input, 'Leia');

    expect(getByDisplayValue('Leia')).toBeDefined();

    // Waiting for the debounce to finish and checking if result element exists and is clickable
    await waitFor(
      () => {
        const resultElement = getByTestId('search-result');
        expect(resultElement).toBeDefined();
        fireEvent.press(resultElement);
      },
      {
        timeout: 2000,
      }
    );

    // Confirming navigation changed the screen to the SwapiCharacter route
    await waitFor(() => {
      expect(getByText('Name:')).toBeDefined();
      expect(getByText('Leia Organa')).toBeDefined();
      expect(getByText('Height:')).toBeDefined();
    });
  });

  it('shows error info when request fails to load', async () => {
    mockedAxios.get.mockRejectedValueOnce('Rejected');
    const { getByText, getByTestId, getByDisplayValue } = setup();

    expect(getByText('Search for the character')).toBeDefined();

    const input = getByTestId('input-text-area');

    expect(input).toBeDefined();

    fireEvent.changeText(input, 'Unknown');

    expect(getByDisplayValue('Unknown')).toBeDefined();

    await waitFor(
      () => {
        const errorElement = getByText('Error occurred, please try again');
        expect(errorElement).toBeDefined();
      },
      {
        timeout: 2000,
      }
    );
  });
});
