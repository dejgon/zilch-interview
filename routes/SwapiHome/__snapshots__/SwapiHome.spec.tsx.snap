// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwapiHome.tsx renders properly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      forwardedRef={[Function]}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backgroundColor="rgb(255, 255, 255)"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="Swapi"
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={false}
      />
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#FFFFFF",
                "flex": 1,
                "padding": 32,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#00D287",
                  "fontFamily": "Oswald-Regular",
                  "fontSize": 64,
                  "fontStyle": "normal",
                  "fontWeight": "400",
                  "lineHeight": 95,
                  "padding": 0,
                }
              }
            >
              Zilch
            </Text>
            <Text
              style={
                {
                  "fontFamily": "Oswald-Regular",
                  "fontSize": 24,
                  "fontStyle": "normal",
                  "fontWeight": "400",
                  "lineHeight": 36,
                }
              }
            >
              Assignment App
            </Text>
            <View
              style={
                {
                  "flex": 1,
                  "gap": 15,
                  "marginTop": 30,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "fontFamily": "Oswald-Regular",
                    "fontSize": 16,
                    "fontStyle": "normal",
                    "fontWeight": "400",
                    "lineHeight": 27,
                    "marginBottom": 10,
                    "textAlign": "center",
                  }
                }
              >
                Find your favourite Star Wars character
              </Text>
              <View
                style={
                  [
                    {
                      "paddingBottom": 16,
                      "width": "100%",
                    },
                    undefined,
                  ]
                }
                testID="input-view-container"
              >
                <Text
                  style={
                    {
                      "color": "#313131",
                      "fontFamily": "Oswald-Regular",
                      "fontSize": 14,
                      "fontStyle": "normal",
                      "fontWeight": "400",
                      "lineHeight": 21,
                      "paddingBottom": 4,
                      "paddingLeft": 4,
                    }
                  }
                  testID="input-label"
                >
                  Search for the character
                </Text>
                <TextInput
                  editable={true}
                  onChangeText={[Function]}
                  style={
                    {
                      "borderColor": "#313131",
                      "borderRadius": 10,
                      "borderStyle": "solid",
                      "borderWidth": 1,
                      "color": "#313131",
                      "fontFamily": "Oswald-Regular",
                      "fontSize": 16,
                      "fontStyle": "normal",
                      "fontWeight": "400",
                      "height": 48,
                      "lineHeight": 24,
                      "paddingLeft": 8,
                      "width": "100%",
                    }
                  }
                  testID="input-text-area"
                  value=""
                />
              </View>
              <Text
                style={
                  {
                    "fontFamily": "Oswald-Light",
                    "fontSize": 16,
                    "fontStyle": "normal",
                    "fontWeight": "400",
                    "lineHeight": 27,
                    "textAlign": "center",
                  }
                }
              >
                No results, try something else ðŸ”Ž
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
